migration/v6/model/ruleset.go

```diff
@@ -0,0 +1,70 @@
+package model
+
+import "gorm.io/gorm"
+
+//
+// RuleSet - Analysis ruleset.
+type RuleSet struct {
+	Model
+	Kind        string
+	Name        string `gorm:"uniqueIndex;not null"`
+	Description string
+	Custom      bool
+	Repository  JSON `gorm:"type:json"`
+	ImageID     uint `gorm:"index" ref:"file"`
+	Image       *File
+	IdentityID  *uint `gorm:"index"`
+	Identity    *Identity
+	Rules       []Rule    `gorm:"constraint:OnDelete:CASCADE"`
+	DependsOn   []RuleSet `gorm:"many2many:RuleSetDependencies;constraint:OnDelete:CASCADE"`
+}
+
+//
+// BeforeUpdate hook to avoid cyclic dependencies.
+func (r *RuleSet) BeforeUpdate(db *gorm.DB) (err error) {
+	var nextDeps []RuleSetDependency
+	var nextRuleSetIDs []uint
+	for _, dep := range r.DependsOn {
+		nextRuleSetIDs = append(nextRuleSetIDs, dep.ID)
+	}
+	for len(nextRuleSetIDs) != 0 {
+		db.Where("RuleSetID IN ?", nextRuleSetIDs).Find(&nextDeps)
+		nextRuleSetIDs = nextRuleSetIDs[:0]
+		for _, nextDep := range nextDeps {
+			if nextDep.DependsOnID == r.ID {
+				err = DependencyCyclicError{}
+				return
+			}
+			nextRuleSetIDs = append(nextRuleSetIDs, nextDep.DependsOnID)
+		}
+	}
+
+	return
+}
+
+//
+// RuleSetDependency represents a ruleset dependency relationship.
+type RuleSetDependency struct {
+	RuleSetID   uint `gorm:"primaryKey"`
+	DependsOnID uint `gorm:"primaryKey"`
+}
```

Comment:
Seems like it would be good to avoid defining the RuleSetDependency model.  curious about why query that table instead of the RuleSet?
Re: BeforeUpdate() - How is Create protected?